### PR TITLE
fixes display other container's process when docker top

### PR DESCRIPTION
**Pull request by lifubang**
Signed-off-by: Lifubang <lifubang@acmcoder.com>

Sometimes, docker top <containerid> [args] may return some process not belonged to my container. This bug will appear in a very small probability.

**- What I did**
~~Remove try ps without -q~~
parse output of ps correctly

**- How I did it**
~~Always with -q~~
find the vertical dividing lines of the output content which is a two-dimensional matrix.
This can also fix the display of the content.

**- How to verify it**
**Before fix:**
**There is a docker entrypoint file: /st**
```
root@dockerdemo:~/gocode# docker exec redisps /bin/cat /st
#!/bin/bash
redis-server --port $1
```
Then start two containers:
```
root@dockerdemo:~/moby# docker run -d --name=redisps --entrypoint=/st redis:testps 6379
root@dockerdemo:~/moby# docker run -d --name=redisps1 --entrypoint=/st redis:testps 17287
root@dockerdemo:~/moby# docker ps | grep redisps
5d19a9c1804b        redis:testps        "/st 17287"         13 minutes ago      Up 13 minutes       6379/tcp            redisps1
4ada0f1dc848        redis:testps        "/st 6379"          2 hours ago         Up 14 minutes       6379/tcp            redisps
root@dockerdemo:~/moby#
```

Docker top:
```
root@dockerdemo:~/moby# docker top redisps
UID                 PID                 PPID                C                   STIME               TTY                 TIME                CMD
root                17287               17271               0                   17:25               ?                   00:00:00            /bin/bash /st 6379
root                17324               17287               0                   17:25               ?                   00:00:00            redis-server *:6379
root@dockerdemo:~/moby# docker top redisps1
UID                 PID                 PPID                C                   STIME               TTY                 TIME                CMD
root                17407               17391               0                   17:25               ?                   00:00:00            /bin/bash /st 17287
root                17439               17407               0                   17:25               ?                   00:00:00            redis-server *:17287
root@dockerdemo:~/moby#
```

The redis's port in container redisps1 is 17287, which is just equals to st's pid in container redisps.
Then run this command:
```
root@dockerdemo:~/moby# docker top redisps -C st -o cmd,user,pid,%cpu
CMD                 USER                PID                 %CPU
/bin/bash           /st                 17287               root 17407 0.0
root@dockerdemo:~/moby# cd /opt/runctest/redis/
```
The PID 17407 belongs to container redisps1, not redisps. If I kill this pid, something will wrong.

**- Description for the changelog**
In most situations, ps -q can work.
such as: ps -ef -q <pids> will return success result.
For some parameters can't work with -q, I think we should not retry without -q, because it is very dangerous.

**- A picture of a cute animal (not mandatory but encouraged)**
xiongmao

### CODE REVIEW

**Comment by thaJeztah** [comment]
Looks like tests are failing;

```
10:57:56 FAIL: docker_api_containers_test.go:435: DockerSuite.TestContainerAPITop
10:57:56 
10:57:56 docker_api_containers_test.go:447:
10:57:56     c.Assert(err, checker.IsNil)
10:57:56 ... value *errors.errorString = &errors.errorString{s:"Error response from daemon: ps: q/-q,--quick-pid cannot be used together with sort options."} ("Error response from daemon: ps: q/-q,--quick-pid cannot be used together with sort options.")
10:57:56 
10:57:56 
```

**Comment by lifubang** [comment]
@thaJeztah, Thanks for your review. If we want each valid ps command can have a result, I think this issue can fix by insuring the first column is pid when user pass -o param to 'docker top' command.

**Comment by kolyshkin** [comment]
> I think we should not retry without -q, because it is very dangerous.

Note `ps -q` optimization was added by commit a076badb8b33f1ecd (PR #37131) and it was not used before it. Also, the fallback to previous mechanism (get info about all processes and apply a filter based on container pids) is there because some `ps` options (than can be specified by a user) conflict with `-q`.

**Comment by kolyshkin** [comment]
Hmm, PR description and commit doesn't match anymore :(

**Comment by kolyshkin** [comment]
OK, it is now clear to be what happens.

1. User specifies `-C <name>` which is incompatible with `-q` used by the optimization, so the old (unoptimized) code runs (list all processes, filter out those that belong to specified container).
2. User specifies `-o cmd,....`, meaning there are some fields before PID that contain spaces. Because of these extra spaces, PID field is not extracted properly.
3. Command executed contains a number which is wrongly recognized as PID. Because of that, some other process (not belonging to this container) is listed.

Now, how to fix this? Changing the order of fields specified by user (or adding a field) looks wrong as `docker top` output will differ from the one expected.

Need to think.

**Comment by kolyshkin** [comment]
We can add PID as the first field, and then remove it from the list.

**Comment by kolyshkin** [comment]
@lifubang your current patch won't work for busybox implementation of `ps` (used by busybox and alpine images, among others), as it does not allow for multiple `-o` arguments. Can you please take a look at https://github.com/moby/moby/pull/37827?

**Comment by lifubang** [comment]
OK, wait for a minute, I am troubled in strings.TrimSpace now, it can't work.

**Comment by kolyshkin** [comment]
@lifubang can you squash the commits into one? Maybe add a commit with my unit tests, too, from https://github.com/moby/moby/pull/37827/commits/09fe7cd4be56b484fb8564f68ec2a3474018a8b7

**Comment by lifubang** [comment]
> @lifubang can you squash the commits into one? Maybe add a commit with my unit tests, too, from [09fe7cd](https://github.com/moby/moby/commit/09fe7cd4be56b484fb8564f68ec2a3474018a8b7)

I don't know how to add your commit with git command, I just create a new file. It will be better if you can tell me how to merge your commit to my branch.

**Comment by lifubang** [comment]
Wait for a minute, there is another bug.
ps o cmd,pid

**Comment by kolyshkin** [comment]
> I don't know how to add your commit with git command

Something like this should work:
```
git add remote kolyshkin https://github.com/kolyshkin/moby
git fetch kolyshkin
git cherry-pick <commitid>
```

Or, faster
```
wget https://github.com/kolyshkin/moby/commit/<commitid>.patch
git am <commitid>.patch
```

**Comment by kolyshkin** [comment]
@lifubang note that since you have added an argument to `parsePSOutput()`, you need to fix its unit test (`daemon/top_unix_test.go`) accordingly (and make sure it passes, just run `cd daemon && go test -v -run TestContainerTop`)

File: daemon/top_unix.go [hunk 1]
```diff
@@ -63,7 +63,7 @@ func hasPid(procs []uint32, pid int) bool {
 	return false
 }
 
-func parsePSOutput(output []byte, procs []uint32) (*container.ContainerTopOKBody, error) {
+func parsePSOutput(output []byte, procs []uint32, blHaso bool) (*container.ContainerTopOKBody, error) {
```
Comment (kolyshkin) [inline daemon/top_unix.go]: I'd suggest to rename `blHaso` to, say, `pidAdded` or something.

File: daemon/top_unix.go [hunk 2]
```diff
@@ -79,6 +79,10 @@ func parsePSOutput(output []byte, procs []uint32) (*container.ContainerTopOKBody
 	if pidIndex == -1 {
 		return nil, fmt.Errorf("Couldn't find PID field in ps output")
 	}
+	//remove appended pid in first cloumn
```
Comment (kolyshkin) [inline daemon/top_unix.go]: * a space after `//` would be nice to have
* `cloumn` -> `column`

File: daemon/top_unix.go [hunk 3]
```diff
@@ -79,6 +79,10 @@ func parsePSOutput(output []byte, procs []uint32) (*container.ContainerTopOKBody
...
+	if blHaso {
+		procList.Titles = procList.Titles[1:]
```
Comment (kolyshkin) [inline daemon/top_unix.go]: You can do something like

At the beginning of the function:

```go
skipCol := 0
if pidAdded {
         skipCol = 1
}
```

and everywhere later in in the code, for example
```go
 	procList.Titles = fieldsASCII(lines[0])[skipCol:]
```
and
```go
	appendProcess2ProcList(procList, fields[skipCol:])
￼```

File: daemon/top_unix.go [hunk 4]
```diff
@@ -130,6 +142,26 @@ func psPidsArg(pids []uint32) string {
 	return string(b)
 }
 
+// insureOPid add a prefix '-o pid' if there is -o in psArgs
```
Comment (kolyshkin) [inline daemon/top_unix.go]: `insure` -> `ensure`
`add` -> `adds`
`a prefix` -> `an option`

File: daemon/top_unix.go [hunk 5]
```diff
@@ -130,6 +142,26 @@ func psPidsArg(pids []uint32) string {
...
+func insureOPid(psArgs string) (string, bool) {
+	var blHaso = false
+	psArr := fieldsASCII(psArgs)
```
Comment (kolyshkin) [inline daemon/top_unix.go]: Note you do `fieldsASCII(psArgs)` here and also in `ContainerTop()`. Perhaps there is no need to do it twice?

File: daemon/top_unix.go [hunk 6]
```diff
@@ -130,6 +142,26 @@ func psPidsArg(pids []uint32) string {
...
+
+	for _, arg := range psArr {
+		if strings.HasPrefix(arg, "-o") || strings.HasPrefix(arg, "--format") || strings.HasPrefix(arg, "o") {
+			blHaso = true
+
```
Comment (kolyshkin) [inline daemon/top_unix.go]: please remove this empty line

**Review by kolyshkin — CHANGES_REQUESTED** [review]

**Comment by kolyshkin** [comment]
@lifubang can you please update my integration test? It appears you're using an old version.

Also, we can remove `validatePSArgs` since we have added PID field ourselves. I think it would be good to check that if the boolean is set, the first field is indeed "PID".

**Comment by lifubang** [comment]
Thank you for your review.

File: daemon/top_unix.go (hunk 1 above)
Comment (lifubang) [reply daemon/top_unix.go]: Thanks.

File: daemon/top_unix.go (hunk 2 above)
Comment (lifubang) [reply daemon/top_unix.go]: Thanks

File: daemon/top_unix.go (hunk 5 above)
Comment (lifubang) [reply daemon/top_unix.go]: No fieldsASCII(psArgs) in ContainerTop

**Comment by lifubang** [comment]
I think my original edition may be useful.
And change -q to -p, it will be safe.

**Comment by lifubang** [comment]
@kolyshkin Maybe use my original solution?

**Comment by lifubang** [comment]
I'm so unhappy that some checks were not successful.
```
10:43:20 PASS: docker_api_swarm_test.go:359: DockerSwarmSuite.TestAPISwarmRaftQuorum	49.392s
10:44:22 
10:44:22 ----------------------------------------------------------------------
10:44:22 FAIL: check_test.go:347: DockerSwarmSuite.TearDownTest
10:44:22 
10:44:22 unmount of /tmp/docker-execroot/d1d88ce69d3dd/netns failed: invalid argument
10:44:22 unmount of /tmp/docker-execroot/d54a7e7608009/netns failed: no such file or directory
10:44:22 check_test.go:352:
10:44:22     d.Stop(c)
10:44:22 /go/src/github.com/docker/docker/internal/test/daemon/daemon.go:402:
10:44:22     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
10:44:22 ... Error: Error while stopping the daemon df8c069be499d : exit status 130
10:44:22 
10:44:22 
10:44:22 ----------------------------------------------------------------------
10:44:22 PANIC: docker_api_swarm_test.go:780: DockerSwarmSuite.TestAPISwarmRestartCluster
```

**Comment by lifubang** [comment]
according to https://salsa.debian.org/debian/procps/blob/master/ps/parser.c
My solution can complete compatible with "ps" command safely. But I can't make CI successful. I think CI errors are not related.

There is another problem. If ps project upgrade, this solution should upgrade too.

**Comment by kolyshkin** [comment]
> 10:43:20 PASS: docker_api_swarm_test.go:359: DockerSwarmSuite.TestAPISwarmRaftQuorum	49.392s

This is a known flaky test, https://github.com/moby/moby/issues/34988 (meaning it is failing randomly and most probably is unrelated to your changes).

**Comment by kolyshkin** [comment]
> 01:23:03.182 FAIL: docker_cli_swarm_test.go:1376: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey

This one is also flaky, https://github.com/moby/moby/issues/33041

**Comment by kolyshkin** [comment]
So, I see the differences between this PR and https://github.com/moby/moby/pull/37827
are quite minimal. Aside from a different lexer, I see that you allow `--no-header` in case of `addPid == true`. Is there anything else I miss @lifubang ?

**Comment by kolyshkin** [comment]
> Aside from a different lexer, I see that you allow --no-header in case of addPid == true.

Sorry, I was wrong, now I see that you are emulating `--no-header`. I don't think it makes sense -- this option was never working before, I doubt we need to add such a feature now.

**Comment by lifubang** [comment]
Yes, @kolyshkin , all main idea is base from you. But I think your lexer can't compatible with all "ps" arg options. I think we need analysis ps arg options with a way like ps source code. I just test my lexer.
For your lexer, -Csto can't work:
```
root@dockerdemo:~# docker top redisps1
UID                 PID                 PPID                C                   STIME               TTY                 TIME                CMD
root                24419               24402               0                   10:43               ?                   00:00:00            /bin/bash /sto 3333
root                24454               24419               0                   10:43               ?                   00:00:00            redis-server *:3333
root@dockerdemo:~# docker top redisps1 -Csto
**error during connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.35/containers/redisps1/top?**ps_args=-Csto: EOF
root@dockerdemo:~# docker top redisps1 -Credis-server
PID                 TTY                 TIME                CMD
24454               ?                   00:00:00            redis-server
```

So, I think you need to upgrade your lexer?

By the way, 

> > Aside from a different lexer, I see that you allow --no-header in case of addPid == true.
> 
> Sorry, I was wrong, now I see that you are emulating `--no-header`. I don't think it makes sense -- this option was never working before, I doubt we need to add such a feature now.

I think --no-header can't working before is also a bug. I think if we can make it work, it will be better.

**Comment by kolyshkin** [comment]
> For your lexer, -Csto can't work:

That's right :( I had it rewritten yesterday, think it handles all the cases now.

**Comment by kolyshkin** [comment]
> I think --no-header can't working before is also a bug. I think if we can make it work, it will be better.

In general, `ContainerTop` is a very bad API, and we should not expand it. I think if it was not working before, let it stay that way.

**Comment by lifubang** [comment]
@kolyshkin For the long term, it should write a new api. For the short term, let's change a way:
To find the vertical dividing line of a two-dimensional matrix.
2cba268c9e3c038f294dfaa8da7e426

**Comment by lifubang** [comment]
For the short term, I think @kolyshkin 's solution may be good enough now. But I still think that we can't list all of the test cases. Such as: PS_FORMAT environment variable, etc. And it's also related to ps project's upgrade.
I think the fault is function named fieldsASCII for this issue. So, I think we may not consider ps options, but consider how to split the output ps command.
So to find the vertical dividing lines of a two-dimensional matrix may be a simple way to fix this issue.
For now, the CI error is for swarm, it's not related to this patch. @thaJeztah @GordonTheTurtle @mlaventure @mlaventure , Please consider it for the short term.

CI Error:
```
12:31:44 ----------------------------------------------------------------------
12:31:44 FAIL: docker_cli_swarm_test.go:274: DockerSwarmSuite.TestSwarmPublishAdd
12:31:44 
12:31:44 [d87a8f9586ef3] waiting for daemon to start
12:31:44 [d87a8f9586ef3] daemon started
12:31:44 
12:31:44 docker_cli_swarm_test.go:286:
12:31:44     c.Assert(err, checker.IsNil, check.Commentf("%s", out))
12:31:44 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc000e0c440), Stderr:[]uint8(nil)} ("exit status 1")
12:31:44 ... Error response from daemon: rpc error: code = Unknown desc = update out of sequence
12:31:44 
12:31:44 
12:31:44 [d87a8f9586ef3] exiting daemon
12:31:49 
12:31:49 ----------------------------------------------------------------------
```

**Comment by kolyshkin** [comment]
> I think the fault is function named fieldsASCII for this issue. So, I think we may not consider ps options, but consider how to split the output ps command.

Yeah, I was thinking about it, too (https://github.com/moby/moby/pull/37827#issuecomment-421521243), but overall we're trying to solve something we should not -- parse the output of a tool that is not supposed to be parsed.

**Comment by olljanat** [comment]
Derek add label: rebuild/*

**Comment by cpuguy83** [comment]
I agree let's not try to parse `ps`.

We could:

1. provide a standard set of data to clients and let clients format
2. allow users to specify a custom command to get the listing/output they want.

I think 1 is more safe, but 2 is more flexible.

In any case, let's close this b/c I think everyone agrees that this is a bad API.
I think it's better to deprecate it than try to fix it.

**Pull request closed without merging**